clash-ghc/src-ghc/Clash/GHC/GHC2Core.hs

```diff
@@ -659,6 +659,8 @@ isSizedCast (TyConApp tc1 _) (TyConApp tc2 _) = do
         ,tc2 `hasKey` naturalTyConKey &&
           tc1Nm == "Clash.Sized.Internal.Unsigned.Unsigned"
         ])
+-- XXX: a work-around for issue #1927, the real fix is to handle all casts
+isSizedCast (TyVarTy {}) (TyConApp tc2 _) = return (tc2 `hasKey` integerTyConKey)
```

Comment:
Does this also need a case for `Integer ~ x`?

Reply:
Not for issue #1927 specifically.

Reply:
Right, but would you trigger the same issue by changing `counter ~ (Index numStages)` to `(Index numStages) ~ counter` in the test? If so I think we should add the extra case to avoid weird (albeit unlikely) situations where someone's design fails unexpectedly unless they change the argument order to `(~)`

Reply:
I haven't tried. I'm somewhat uncomfortable with this PR to begin with. I'd rather properly support all casts.

Reply:
True, supporting all casts would be the better option. How far was #1064 from being finished?

Reply:
It went in the right direction. But I would want to redo some aspects, especially with regards to `Tick`. So [`squashCollectApp`](https://github.com/clash-lang/clash-compiler/compare/castSignal-rebased#diff-1051fc7ced5b1944b216dd30a323e3c1fd9410a615136ff4c717089c899888c8R608-R716) was there in terms of handling casts I think, but `Ticks` still need more care: depending on the type of info (source location, HDL naming, etc.) the ticks either should or shouldn't be distributed towards the arguments as we "squash" the entire application.